### DISCUSSION TITLE
Revenir à une API fonctionnelle basique

### DISCUSSION THREAD

**Discussion by mquandalle** [publicodes/publicodes#260] [General]
J'ai bataillé quelques heures contre des problèmes de réactivité entre Svelte et l'API impérative du moteur Publicodes. Le [code résultant](https://github.com/mquandalle/mesaidesvelo/commit/4fe4b24dae3f6a0b8d7a9e4e2388035c5eeddb22) est moche, sans doute buggé, et repose sur des `engineBis = engine.shallowCopy()` et `.setSituation({...currentSituation, modif})` dispersés un peu partout sans grande logique. 

Ce n'est pas la première fois que je rencontre ces soucis : le moteur est facile à utiliser si on arrive à entrer tous les calculs directement dans la base de règles publicodes et qu'il suffit d'un `omnipotentEngineInstance.evaluate(rule)` pour obtenir le résultat voulu. Mais si on a besoin de relancer plusieurs calculs dans plusieurs situations amendées de manière réactive, la chose devient beaucoup plus ardue — et quasi-impossible si en plus on veut le faire de manière performante sans invalidation de cache inutile.

Peut-être est-ce un mirage mais on pourrait revenir à une API beaucoup plus simple et fonctionnelle avec essentiellement : 
- une fonction `parse(jsonSource)` (utilisable côté serveur cf. #256)
- une fonction `evaluate(parsedRules, ruleName)`

C'était un peu l'API qu'on avait initialement avant [l'introduction d'un objet JS](https://github.com/betagouv/mon-entreprise/issues/743). L'une des motivation de l'API objet était la gestion d'un état interne, et en particulier `this.cache` qui permet de ne pas recalculer 2 fois le même nœud. 

Mais il me semble qu'un cache global de type `WeakMap` qui permet d'associer une valeur calculée à un nœud immuable donné en entrée, rend possible de gérer un cache en gardant une API fonctionnelle type `evaluate(rules, ruleName)` et donc de supprimer `.shallowCopy()`. Je pense même que le concept de situation pourrait être sorti du moteur, puisqu'en soit le calcul avec situation revient à faire :

```js
evaluate({ ...baseRules, ...parsedSituation }, ruleName)
```

Notamment la chose que je n'avais pas réalisé est que le spread operator de JS conserve l'identité des objets, ce qui veut dire que tous les calculs cachés dans une `WeakMap` n'ont pas besoin d'être refaits même si les paramètres donnés en entrée sont modifiés :

```js
const baseRules = { a : {} }, amendedRules = { ...baseRules }
console.assert(baseRules.a === amendedRules.a)
```

Bref évidemment ça pose pas mal de questions pour l’implementation, mais je crois que la réflexion vaut le coup. Vous avez déjà rencontré aussi ce type de difficulté avec l'API impérative ? Vous voyez des blocages pour implémenter une API fonctionnelle plus minimaliste ?

**Comment by laem** [comment]
Pas trop de souci de mon côté. Mais je crois comprendre ton point. J'aime bien l'idée de l'API plus simple, mais en effet, tout est question de cache.
